### PR TITLE
IGNITE-17321 Add docs for java client partition awareness use cases

### PR DESCRIPTION
**Pull request by J-Bakuli**
In javadoc org.apache.ignite.configuration.ClientConfiguration#partitionAwarenessEnabled and in the description of functionality
https://ignite.apache.org/docs/latest/thin-clients/java-thin-client#partition-awareness, it is not described with which api this functionality will work and in what cases. For example, will it work with getAll, in a transaction?

Describe in the documentation and in the javadoc in which cases it works and with which api.
______________
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: docs/_docs/thin-clients/java-thin-client.adoc [hunk 1]
```diff
@@ -79,6 +79,10 @@ Note that the code above provides a failover mechanism in case of server node fa
 
 include::includes/partition-awareness.adoc[]
 
+Partition awareness functionality helps to avoid an additional network hop in the following scenarios:
+1. Single-key operations API, like #put, #get, etc. However, the functionality has no effect on those operations within explicit transactions (initiated via ClientTransaction.txStart() described in <<Executed transactions>> section).
```
Comment (timoninmaxim) [inline docs/_docs/thin-clients/java-thin-client.adoc:83]: I set up docs locally and link to "Executing Scan Queries" works, but links on transactions and IndexQueries doesn't. Could you, please, check it?

File: docs/_docs/code-snippets/java/src/main/java/org/apache/ignite/snippets/JavaThinClient.java [hunk 2]
```diff
@@ -359,6 +359,9 @@ void partitionAwareness() throws Exception {
         try (IgniteClient client = Ignition.startClient(cfg)) {
             ClientCache<Integer, String> cache = client.cache("myCache");
             // Put, get or remove data from the cache...
+            cache.put(0; "Hello, world!");
```
Comment (timoninmaxim) [inline docs/_docs/code-snippets/java/src/main/java/org/apache/ignite/snippets/JavaThinClient.java:362]: replace ';' with ','

File: docs/_docs/thin-clients/java-thin-client.adoc [hunk 3]
```diff
@@ -79,6 +79,10 @@ Note that the code above provides a failover mechanism in case of server node fa
...
+2. ScanQuery and IndexQuery accept a partition number as a parameter with which the query is routed to a particular server node that stores the requested data. Refer to <<Executing Scan Queries>> and <<Executing Index Queries>> sections for more information.
```
Comment (timoninmaxim) [inline docs/_docs/thin-clients/java-thin-client.adoc:84]: Also, you should mark numerations correctly - to enforce every point starts  from new line. Currently, it is showed in a single line.

File: docs/_docs/thin-clients/java-thin-client.adoc (hunk 1 above)
Comment (timoninmaxim) [inline docs/_docs/thin-clients/java-thin-client.adoc:83]: "#put, #get" -> `put(), get()`

**Review by timoninmaxim — CHANGES_REQUESTED** [review]

File: docs/_docs/thin-clients/java-thin-client.adoc (hunk 1 above)
Comment (J-Bakuli) [reply docs/_docs/thin-clients/java-thin-client.adoc]: I've corrected the links to Transactions and IndexQueries
Thank you

File: docs/_docs/thin-clients/java-thin-client.adoc [hunk 4]
```diff
@@ -79,6 +79,12 @@ Note that the code above provides a failover mechanism in case of server node fa
 
 include::includes/partition-awareness.adoc[]
 
+Partition awareness functionality helps to avoid an additional network hop in the following scenarios:
+
+1. Single-key operations API, like put(), get(), etc. However, the functionality has no effect on those operations within explicit transactions (initiated via ClientTransaction.txStart() described in <<Transactions>> section).
+
+2. ScanQuery and IndexQuery accept a partition number as a parameter with which the query is routed to a particular server node that stores the requested data. Refer to <<Executing Scan Queries>> and link:key-value-api/using-cache-queries[Executing Index Queries] sections for more information.
```
[the remark this answers is not in the document]
Comment (J-Bakuli) [reply docs/_docs/thin-clients/java-thin-client.adoc]: thank you

File: modules/core/src/main/java/org/apache/ignite/configuration/ClientConfiguration.java [hunk 5]
```diff
@@ -529,6 +531,11 @@ public ClientConfiguration setTransactionConfiguration(ClientTransactionConfigur
     }
 
     /**
+     * Partition awareness functionality helps to avoid an additional network hop in the following scenarios:
+     * 1. Single-key operations API, like put(), get(), etc. However, the functionality has no effect on those operations
```
[the remark this answers is not in the document]
Comment (J-Bakuli) [reply modules/core/src/main/java/org/apache/ignite/configuration/ClientConfiguration.java]: thank you, I have corrected

File: modules/core/src/main/java/org/apache/ignite/configuration/ClientConfiguration.java [hunk 6]
```diff
@@ -529,11 +531,17 @@ public ClientConfiguration setTransactionConfiguration(ClientTransactionConfigur
     }
 
     /**
+     * Partition awareness functionality helps to avoid an additional network hop in the following scenarios:
+     * <ul>
+     *     <li>1. Single-key operations API, like put(), get(), etc. However, the functionality has no effect on those
+     *     operations within explicit transactions {@code ClientTransactions#txStart()}.</li>
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/configuration/ClientConfiguration.java:537]: here and below, let's replace @code with @link, and then import required dependencies, to make it possible to navigate for mentioned functions

File: modules/core/src/main/java/org/apache/ignite/configuration/ClientConfiguration.java [hunk 7]
```diff
@@ -529,11 +531,17 @@ public ClientConfiguration setTransactionConfiguration(ClientTransactionConfigur
...
+     *     <li>2. {@link ScanQuery#setPartition(Integer)} and {@link IndexQuery#setPartition(Integer)} accept a
+     *     partition number as a parameter with which the query is routed to a particular server node that stores
+     *     the requested data.</li>
+     * </ul>
      * @return A value indicating whether partition awareness should be enabled.
-     * <p>
-     * Default is {@code true}: client sends requests directly to the primary node for the given cache key.
-     * To do so, connection is established to every known server node.
-     * <p>
+     * <p>Default is {@code true}: client sends requests directly to the primary node for the given cache key.
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/configuration/ClientConfiguration.java:543]: Let's move this part at the beginning of javadoc. @return should be the last line in the doc:
```
<p>Default is  ....
When {@code false}...
<p>
Partition awareness functionality...

```

File: docs/_docs/thin-clients/java-thin-client.adoc [hunk 8]
```diff
@@ -79,6 +79,12 @@ Note that the code above provides a failover mechanism in case of server node fa
 
 include::includes/partition-awareness.adoc[]
 
+Partition awareness functionality helps to avoid an additional network hop in the following scenarios:
+
+1. Single-key operations API, like put(), get(), etc. However, the functionality has no effect on those operations within explicit transactions (initiated via ClientTransaction.txStart() described in <<Transactions>> section).
```
Comment (timoninmaxim) [inline docs/_docs/thin-clients/java-thin-client.adoc:84]: .txStart() -> #txStart()

File: docs/_docs/code-snippets/java/src/main/java/org/apache/ignite/snippets/JavaThinClient.java [hunk 9]
```diff
@@ -359,6 +359,9 @@ void partitionAwareness() throws Exception {
         try (IgniteClient client = Ignition.startClient(cfg)) {
             ClientCache<Integer, String> cache = client.cache("myCache");
             // Put, get or remove data from the cache...
+            cache.put(0, "Hello, world!");
+            // The partition number can be specified with indexQuery.setPartition(Integer) as well.
```
Comment (timoninmaxim) [inline docs/_docs/code-snippets/java/src/main/java/org/apache/ignite/snippets/JavaThinClient.java:363]: indexQuery.setPartition(Integer) -> IndexQuery#setPartition(Integer)

**Review by timoninmaxim — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/configuration/ClientConfiguration.java (hunk 6 above)
Comment (J-Bakuli) [reply modules/core/src/main/java/org/apache/ignite/configuration/ClientConfiguration.java:537]: thank you, I will correct

File: modules/core/src/main/java/org/apache/ignite/configuration/ClientConfiguration.java (hunk 7 above)
Comment (J-Bakuli) [reply modules/core/src/main/java/org/apache/ignite/configuration/ClientConfiguration.java:543]: thank you, I will correct

**Review by timoninmaxim — APPROVED** [review]

**Pull request merged**